### ISSUE TITLE
Nesting level too deep - recursive dependency? when using block.getPrev.type == block.type

### ISSUE THREAD

**Issue by mahajanl** [craftcms/cms#6759]
### Description
When attempting to use the following code:
```twig
{% if block.getPrev.type == block.type %}
  <div class="grid grid-2">
{% endif %}
```

i get the following error: 

> Nesting level too deep - recursive dependency?

### Steps to reproduce

1. create a matrix field with more than one block type and add it to a section; give it a few blocks.
2. attempt to template and use `{% if block.getPrev.type == block.type %}`
3. a note - i can get `block.getPrev.type` and `block.type` just fine, but cannot check if theyre the same, even after setting them as variables. I have definitely used this code a number of times in the past. 

### Additional info

- Craft version: 3.5.7
- PHP version: 7.2
- Database driver & version: 
- Plugins & versions: 
  Asset Rev | 6.0.2
  Control Panel Nav 3.0.13.1
  Craft Commerce 3.2.3
  Element API 2.6.0
  Feed Me 4.2.3
  oEmbed 1.3.4
  Redactor 2.7.3
  Sprig 1.0.0-beta.11
  Template Comments 1.1.2

thanks!

**Comment by brandonkelly** [comment]
I’m not able to reproduce that error; the code works on my end. Can you please send your Composer files and a database backup over to support@craftcms.com? We can try to look into it from there.

**Comment by angrybrad** [comment]
@mahajanl any change you've got xDebug installed?

**Comment by mahajanl** [comment]
Shall do, thanks!

Lakshami Mahajan
UI
 Developer
for The Refinery <http://the-refinery.io>
Freelance Developer

**Comment by mahajanl** [comment]
i do not. but i can if it will help.

Lakshami Mahajan
UI
 Developer
for The Refinery <http://the-refinery.io>
Freelance Developer

**Comment by mahajanl** [comment]
i've had another dev on our team reproduce it on his 3.5.5 install as well.
perhaps i havent explained the issue enough though, let me try again.

I have a matrix field with a number of blocks. the field is pagebuilder;
this is the template code for sections with that field

{% for block in entry.pagebuilder.all() %}
  {% include '_blocks/' ~ block.type %}
{% endfor %}

and then the offending bit in the block partial is

{% if block.getPrev.type == block.type %}

even when I ensure it's not loop.first

hope this helps, and thanks again.



Lakshami Mahajan
UI
 Developer
for The Refinery <http://the-refinery.io>
Freelance Developer

**Comment by mahajanl** [comment]
UPDATE: a fellow dev found the issue: It's the php version. I'm on 7.2. 7.3 and 7.4 are fine.

**Comment by brandonkelly** [comment]
Glad you got it sorted!

**Comment by mahajanl** [comment]
its not really sorted - i can’t upgrade our server at the moment (i did it
locally just to see if that was the fix), and anyone else still running 7.2
would run into this issue - including some upgrades we have in progress
from 3.4.

Lakshami Mahajan
UI
 Developer
for The Refinery <http://the-refinery.io>
Freelance Developer

**Comment by mtwalsh** [comment]
Just adding a note here in case it saves someone else some time...

I hit this error on Craft `4.7.2` and PHP `8.2.15` but was able to resolve it by changing the comparison operator:

So this:

`{% if previousBlockType != thisBlockType %}`

Becomes this:

`{% if previousBlockType is not same as(thisBlockType) %}`

I've also seen this error on another project, also PHP `8.2.15`, when the `|sort` filter was used, removing that and doing the sort at the query level (makes a lot of sense) resolved the issue.

**Comment by brandonkelly** [comment]
@mtwalsh If you can send a database backup and your Composer files into support@craftcms.com, along with example template code that we could use to replicate, we can look into it from there.
